src/sentry/sentry_metrics/indexer/limiters/cardinality.py

```diff
@@ -0,0 +1,148 @@
+from __future__ import annotations
+
+import dataclasses
+from collections import defaultdict
+from typing import Any, Generic, Mapping, MutableMapping, Optional, Sequence, TypedDict, TypeVar
+
+from sentry import options
+from sentry.ratelimits.cardinality import (
+    CardinalityLimiter,
+    GrantedQuota,
+    Quota,
+    RedisCardinalityLimiter,
+    RequestedQuota,
+    Timestamp,
+)
+from sentry.sentry_metrics.configuration import MetricsIngestConfiguration, UseCaseKey
+from sentry.utils import metrics
+from sentry.utils.hashlib import hash_values
+
+OrgId = int
+
+TMessageKey = TypeVar("TMessageKey")
+
+
+@dataclasses.dataclass(frozen=True)
+class CardinalityLimiterState(Generic[TMessageKey]):
+    _cardinality_limiter: CardinalityLimiter
+    _use_case_id: UseCaseKey
+    _grants: Sequence[GrantedQuota]
+    _timestamp: Timestamp
+    keys_to_remove: Sequence[TMessageKey]
+
+
+def _build_quota_key(namespace: str, org_id: Optional[OrgId]) -> str:
+    if org_id is not None:
+        return f"metrics-indexer-cardinality-{namespace}-org-{org_id}"
+    else:
+        return f"metrics-indexer-cardinality-{namespace}-global"
+
+
+@metrics.wraps("sentry_metrics.indexer.construct_quotas")
+def _construct_quotas(use_case_id: UseCaseKey) -> Sequence[Quota]:
+    """
+    Construct write limit's quotas based on current sentry options.
+
+    This value can potentially cached globally as long as it is invalidated
+    when sentry.options are.
+    """
+    if use_case_id == UseCaseKey.PERFORMANCE:
+        return [
+            Quota(**args)
+            for args in options.get("sentry-metrics.writes-limiter.limits.performance.per-org")
+        ]
+    elif use_case_id == UseCaseKey.RELEASE_HEALTH:
+        return [
+            Quota(**args)
+            for args in options.get("sentry-metrics.writes-limiter.limits.releasehealth.per-org")
+        ]
+    else:
+        raise ValueError(use_case_id)
+
+
+class InboundMessage(TypedDict):
+    # Note: This is only the subset of fields we access in this file.
+    org_id: int
+    name: str
+    type: str
+    tags: Mapping[str, str]
+
+
+class TimeseriesCardinalityLimiter:
```

Comment:
Please add a unit test. There is enough logic in this class that is granted a test even if the rate limiter does not do anything.

Reply:
added some basic tests, found some pretty severe bugs 💯